modules/sdk-lib-mpc/src/tss/ecdsa-dkls/dsg.ts

```diff
@@ -0,0 +1,169 @@
+import { SignSession, Keyshare, Message } from '@silencelaboratories/dkls-wasm-ll-node';
+import { DeserializedBroadcastMessage, DeserializedMessages, DklsSignature, DsgState } from './types';
+import { decode } from 'cbor';
+
+export class Dsg {
+  protected dsgSession: SignSession | undefined;
+  protected dsgSessionBytes: Uint8Array;
+  protected signature: DklsSignature | undefined;
+  protected keyShare: Keyshare;
+  protected messageHash: Buffer;
+  protected derivationPath: string;
+  protected partyIdx: number;
+  protected dsgState: DsgState = DsgState.Uninitialized;
+
+  constructor(keyShare: Buffer, partyIdx: number, derivationPath: string, messageHash: Buffer) {
+    this.keyShare = Keyshare.fromBytes(keyShare);
+    this.partyIdx = partyIdx;
+    if (this.keyShare.partyId !== partyIdx) {
+      throw Error(`Party index: ${partyIdx} does not match key share partyId: ${this.keyShare.partyId} `);
+    }
+    this.derivationPath = derivationPath;
+    this.messageHash = messageHash;
+  }
+
+  private _restoreSession() {
+    if (!this.dsgSession) {
+      this.dsgSession = SignSession.fromBytes(this.dsgSessionBytes);
+    }
+  }
+
+  private _deserializeState() {
+    if (!this.dsgSession) {
+      throw Error('Session not intialized');
+    }
+    const round = decode(this.dsgSession.toBytes()).round;
```

Comment:
This is something that we may want to bring up with SL - to have a `getRound()` function  so you don't have to do this.

Reply:
Nit: pass the result of `decode` into an io-ts codec. Invalid object shapes would result in a runtime error, and also give us type-safe objects to work with.

Reply:
i think having the SL code have a statically type defined function getRound will be better, hence will communicate that with their team. Will add a codec in a separate PR if they dont end up creating a getRound function. Ticket to track https://bitgoinc.atlassian.net/browse/HSM-315